subsys/net/lib/lwm2m/lwm2m_obj_access_control.c

```diff
@@ -0,0 +1,433 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_MODULE_NAME net_lwm2m_ac_control
+#define LOG_LEVEL	CONFIG_LWM2M_LOG_LEVEL
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include "lwm2m_obj_access_control.h"
+
+#include <stdint.h>
+
+#include <zephyr/init.h>
+
+#define ACCESS_CONTROL_VERSION_MAJOR 1
+#define ACCESS_CONTROL_VERSION_MINOR 0
+
+/* Local, for readability */
+#define READ   ACCESS_CONTROL_PERM_READ
+#define WRITE  ACCESS_CONTROL_PERM_WRITE
+#define ACEXEC   ACCESS_CONTROL_PERM_EXEC
+#define DELETE ACCESS_CONTROL_PERM_DELETE
+#define CREATE ACCESS_CONTROL_PERM_CREATE
+
+/* For compatibility with lwm2m_op_perms */
+#define WRITE_ATTR BIT(8)
+#define DISCOVER   BIT(9)
+
+static int lwm2mperm_to_acperm(int lwm2mperm)
+{
+	switch (lwm2mperm) {
+	case LWM2M_OP_READ:
+		return READ;
+
+	case LWM2M_OP_WRITE:
+		return WRITE;
+
+	case LWM2M_OP_EXECUTE:
+		return ACEXEC;
+
+	case LWM2M_OP_DELETE:
+		return DELETE;
+
+	case LWM2M_OP_CREATE:
+		return CREATE;
+
+	case LWM2M_OP_WRITE_ATTR:
+		return WRITE_ATTR;
+
+	case LWM2M_OP_DISCOVER:
+		return DISCOVER;
+	default:
+		return 0;
+	}
+}
+
+#define AC_OBJ_ID		    LWM2M_OBJECT_ACCESS_CONTROL_ID
+
+#define ACCESS_CONTROL_OBJECT_ID	    0
+#define ACCESS_CONTROL_OBJECT_INSTANCE_ID   1
+#define ACCESS_CONTROL_ACL_ID		    2
+#define ACCESS_CONTROL_ACCESS_CONTROL_OWNER 3
+#define ACCESS_CONTROL_MAX_ID		    4
+
+#define MAX_SERVER_COUNT   CONFIG_LWM2M_SERVER_INSTANCE_COUNT
+#define MAX_INSTANCE_COUNT 60 /* TODO: figure out this number. Max # of ac object instances */
```

Comment:
IMO it'd make sense to have a Kconfig entry for this parameter, with some arbitrary default. The application could then adjust the instance number per its needs.